src/ophys_etl/modules/segmentation/seed/seeder.py

```diff
@@ -0,0 +1,317 @@
+from scipy.ndimage import gaussian_filter
+from scipy.spatial.distance import cdist
+
+from ophys_etl.modules.segmentation.seed.utils import (Seed,
+                                                       dilated_coordinates)
+
+
+class SeederBase(Iterator):
+    """generic base class for seeding, and supports HNCcorr style of seeder
+
+    Parameters
+    ----------
+    fov_shape: Tuple[int, int]
+        the (nrows, ncols) shape of the field-of-view. Used by
+        'dilated_coordinates()` to restrict excluded coordinates to
+        a certain FOV
+    exclusion_buffer: int
+        the number of pixels beyond an ROI (by binary dilation) for ignoring
+        subsequent seeds.
+
+    Attributes
+    ----------
+    _candidate_seeds: List[Seed]:
+        the list of seeds created by 'select_seeds'
+    _provided_seeds: List[Seed]:
+        the list of seeds that have been served by the iterator
+    _excluded_seeds: List[Seed]:
+        the list of candidate seeds that were not served by the iterator
+    _excluded_pixels: Set[Tuple[Int, Int]]
+        list of current excluded pixels, excluded by 'exclude_pixels()'
+    _exclusion_buffer: int
+        see Parameter 'exclusion_buffer'
+
+    Notes
+    -----
+    The expected use of this class's children:
+    ```
+    seeder = SeederChildClass(*args)
+    segmenter = SegmentClass(seeder)
+
+    # segmenter uses seeder as iterator,updating exclusions
+    segmenter.run()
+
+    with h5py.File("segmentation_log.h5", "w") as f:
+        segmenter.log_to_h5_group(f, "segmenting")
+
+        # log seeder after segmenter has finished using it
+        seeder.log_to_h5_group(f, "seeding")
+    ```
+
+    """
+    def __init__(self,
+                 exclusion_buffer: int = 1):
+        self._candidate_seeds: List[Seed] = []
+        self._provided_seeds: List[Seed] = []
+        self._excluded_seeds: List[Seed] = []
+        self._exclusion_buffer = exclusion_buffer
+        self._excluded_pixels: Set[Tuple[int, int]] = set()
+
+    def exclude_pixels(self, pixels):
+        """applies dilation with a 3x3 (default) square structuring element
+        and adds new pixels to self._excluded_pixels
+
+        Parameters
+        ----------
+        pixels (set): Set of pixel coordinates to exclude.
+
+        Returns
+        -------
+        None
+
+        """
+        dilated_pixels = dilated_coordinates(pixels,
+                                             self._exclusion_buffer)
+        self._excluded_pixels.update(dilated_pixels)
+
+    def __next__(self):
+        """returns the next valid seed
+
+        Notes
+        -----
+        A seed can be invalid either because it was too low on the list of
+        seeds sorted by metric (excluded_by_fraction) or if it was within
+        a buffer of an ROI and is in _excluded_pixels (excluded_by_roi)
+
+        """
+        if len(self._candidate_seeds) == 0:
+            raise StopIteration
+
+        next_seed = self._candidate_seeds.pop(0)
```

Comment:
This is a nitpick but, self._candidate_seeds could probably be a [`collections deque`](https://docs.python.org/3/library/collections.html#collections.deque). Popping the 0-th element from a normal `list` is an O(N) operation since all subsequent list elements need to be shifted.

Popping the 0-th element of a `deque` is amortized O(1).

Reply:
That is interesting, thanks for pointing that out.
But, a subclass of this, `BatchImageMetricSeeder` has and needs the ability to `pop` from the middle of the list. I don't think `deque` supports this.
The scenario where this happens in `BatchImageMetricSeeder` is: the top 2 seeds are within some minimum distance of each other. The 2nd seed is passed over, kept in `_candidates` until the next iteration. But, to fill out the first batch, the 3rd seed might be chosen.